Source/WebCore/Modules/url-pattern/URLPattern.h

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright (C) 2024 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <wtf/Forward.h>
+#include <wtf/Ref.h>
+#include <wtf/RefCounted.h>
+#include <wtf/RefPtr.h>
+#include <wtf/TZoneMalloc.h>
+
+namespace WebCore {
+
+struct URLPatternInit;
+struct URLPatternOptions;
+struct URLPatternResult;
+
+class URLPattern final : public RefCounted<URLPattern> {
+    WTF_MAKE_TZONE_OR_ISO_ALLOCATED(URLPattern);
+public:
+    using URLPatternInput = std::variant<String, URLPatternInit>;
+
+    static Ref<URLPattern> create(const URLPatternInput, const String&& baseURL, URLPatternOptions&&);
+    static Ref<URLPattern> create(std::optional<URLPatternInput>, URLPatternOptions&&);
+    ~URLPattern();
+
+    ExceptionOr<bool> test(std::optional<URLPatternInput>, std::optional<String>&& baseURL) const;
+
+    std::optional<URLPatternResult> exec(std::optional<URLPatternInput>&&, std::optional<String>&& baseURL) const;
```

Comment:
Does it build if we use String&& for the second parameter? String has a null state in WebKit (String::isNull()) so we generally do not need to use std::optional<String>